File: src/csharp/howto/HowTo_ECMO.cs
/* Distributed under the Apache License, Version 2.0.
   See accompanying NOTICE file for details.*/

using System;
using System.Collections.Generic;
using Pulse;
using Pulse.CDM;

namespace HowTo_ECMO
{
  class Example
  {
    public static void Run()
    {
      // Create our pulse engine
      PulseEngine pulse = new PulseEngine();
      pulse.LogToConsole(true);// Easily view what is happening

      List<SEDataRequest> data_requests = new List<SEDataRequest>
      {
        // Vitals Monitor Data
        SEDataRequest.CreatePhysiologyDataRequest("RespirationRate", FrequencyUnit.Per_min),
        SEDataRequest.CreatePhysiologyDataRequest("TidalVolume", VolumeUnit.mL),
        SEDataRequest.CreatePhysiologyDataRequest("TotalLungVolume", VolumeUnit.mL),
        SEDataRequest.CreatePhysiologyDataRequest("ExpiratoryPulmonaryResistance", PressureTimePerVolumeUnit.cmH2O_s_Per_L),
        SEDataRequest.CreatePhysiologyDataRequest("InspiratoryPulmonaryResistance", PressureTimePerVolumeUnit.cmH2O_s_Per_L),
        SEDataRequest.CreatePhysiologyDataRequest("PulmonaryCompliance", VolumePerPressureUnit.L_Per_cmH2O),
        SEDataRequest.CreatePhysiologyDataRequest("TotalPulmonaryVentilation", VolumePerTimeUnit.L_Per_min),
        SEDataRequest.CreateLiquidCompartmentDataRequest("ECMOBloodSamplingPort", "InFlow"),
        SEDataRequest.CreateLiquidCompartmentDataRequest("ECMOBloodSamplingPort", "OutFlow"),
        SEDataRequest.CreateLiquidCompartmentDataRequest("ECMOOxygenator", "InFlow"),
        SEDataRequest.CreateLiquidCompartmentDataRequest("ECMOOxygenator", "OutFlow"),
        SEDataRequest.CreateLiquidCompartmentDataRequest("RightArmVasculature", "InFlow"),
        SEDataRequest.CreateLiquidCompartmentDataRequest("RightArmVasculature", "OutFlow"),
      };
      SEDataRequestManager data_mgr = new SEDataRequestManager(data_requests);
      data_mgr.SetResultsFilename("./test_results/howto/HowTo_ECMO.cs.csv");
      // Create a reference to a double[] that will contain the data returned from Pulse
      double[] data_values;
      // data_values[0] is ALWAYS the simulation time in seconds
      // The rest of the data values are in order of the data_requests list provided

      if (!pulse.SerializeFromFile("./states/StandardMale@0s.pbb", data_mgr))
      {
        Console.WriteLine("Error Initializing Pulse!");
        return;
      }
      pulse.AdvanceTime_s(10);
      // Get the values of the data you requested at this time
      data_values = pulse.PullData();
      // And write it out to the console
      data_mgr.WriteData(data_values);

      // Give the patient Dyspnea
      SEECMOConfiguration cfg = new SEECMOConfiguration();
      SEECMOSettings settings = cfg.GetSettings();
      settings.SetInflowLocation(eECMO_CannulationLocation.InternalJugular);
      settings.SetOutflowLocation(eECMO_CannulationLocation.RightSubclavianVein);
      settings.GetOxygenatorVolume().SetValue(500, VolumeUnit.mL);
      settings.GetTransfusionFlow().SetValue(1.0, VolumePerTimeUnit.mL_Per_s);
      settings.SetSubstanceCompound("Saline");
      pulse.ProcessAction(cfg);

      pulse.AdvanceTime_s(10);
      // Get the values of the data you requested at this time
      data_values = pulse.PullData();
      // And write it out to the console
      data_mgr.WriteData(data_values);

    }
  }
}

File: src/java/com/kitware/pulse/howto/HowTo_ECMO.java
/* Distributed under the Apache License, Version 2.0.
   See accompanying NOTICE file for details.*/
package com.kitware.pulse.howto;

import java.util.List;

import com.kitware.pulse.cdm.bind.ECMO.eECMOCannulationLocation;
import com.kitware.pulse.cdm.engine.SEDataRequestManager;
import com.kitware.pulse.cdm.properties.CommonUnits.*;
import com.kitware.pulse.cdm.system.equipment.ecmo.SEECMOSettings;
import com.kitware.pulse.cdm.system.equipment.ecmo.actions.SEECMOConfiguration;
import com.kitware.pulse.engine.PulseEngine;
import com.kitware.pulse.utilities.JNIBridge;
import com.kitware.pulse.utilities.Log;
import com.kitware.pulse.utilities.LogListener;

public class HowTo_ECMO
{
  protected static class MyListener extends LogListener
  {
    protected MyListener()
    {
      super();
      listen(false);
    }    
    @Override public void handleDebug(String msg) { Log.debug(msg); }
    @Override public void handleInfo(String msg)  { Log.info(msg); }
    @Override public void handleWarn(String msg)  { Log.warn(msg); }
    @Override public void handleError(String msg) { Log.error(msg); }
    @Override public void handleFatal(String msg) { Log.fatal(msg); }
  }
  
  public static void main(String[] args)
  {
    List<Double> data_values;
    
    JNIBridge.initialize();
    PulseEngine pulse = new PulseEngine();
    pulse.setLogListener(new MyListener());
    pulse.setLogFilename("./test_results/howto/HowTo_ECMO.java.log");
    
    SEDataRequestManager dataRequests = new SEDataRequestManager();// Vitals Monitor Data
    dataRequests.createPhysiologyDataRequest("HeartRate", FrequencyUnit.Per_min);
    dataRequests.createPhysiologyDataRequest("ArterialPressure", PressureUnit.mmHg);
    dataRequests.createPhysiologyDataRequest("MeanArterialPressure", PressureUnit.mmHg);
    dataRequests.createPhysiologyDataRequest("SystolicArterialPressure", PressureUnit.mmHg);
    dataRequests.createPhysiologyDataRequest("DiastolicArterialPressure", PressureUnit.mmHg);
    dataRequests.createPhysiologyDataRequest("OxygenSaturation");
    dataRequests.createPhysiologyDataRequest("EndTidalCarbonDioxidePressure", PressureUnit.mmHg);
    dataRequests.createPhysiologyDataRequest("RespirationRate", FrequencyUnit.Per_min);
    dataRequests.createPhysiologyDataRequest("SkinTemperature", TemperatureUnit.C);
    dataRequests.createGasCompartmentDataRequest("Carina", "CarbonDioxide", "PartialPressure", PressureUnit.mmHg);
    dataRequests.createPhysiologyDataRequest("BloodVolume", VolumeUnit.mL);
    dataRequests.createECGDataRequest("Lead3ElectricPotential", ElectricPotentialUnit.mV);
    // Equipment Data
    dataRequests.createLiquidCompartmentDataRequest("ECMOBloodSamplingPort", "InFlow");
    dataRequests.createLiquidCompartmentDataRequest("ECMOBloodSamplingPort", "OutFlow");
    dataRequests.createLiquidCompartmentDataRequest("ECMOOxygenator", "InFlow");
    dataRequests.createLiquidCompartmentDataRequest("ECMOOxygenator", "OutFlow");
    dataRequests.createLiquidCompartmentDataRequest("RightArmVasculature", "InFlow");
    dataRequests.createLiquidCompartmentDataRequest("RightArmVasculature", "OutFlow");
    
    dataRequests.setResultsFilename("./test_results/howto/HowTo_ECMO.java.csv");
    
    pulse.serializeFromFile("./states/StandardMale@0s.json", dataRequests);
    
    pulse.advanceTime_s(10);
    // Get the values of the data you requested at this time
    data_values = pulse.pullData();
    // And write it out to the console
    dataRequests.writeData(data_values);

    SEECMOConfiguration cfg = new SEECMOConfiguration();
    SEECMOSettings settings = cfg.getSettings();
    settings.setInflowLocation(eECMOCannulationLocation.InternalJugular);
    settings.setOutflowLocation(eECMOCannulationLocation.RightSubclavianVein);
    settings.getOxygenatorVolume().setValue(500, VolumeUnit.mL);
    settings.getTransfusionFlow().setValue(1.0, VolumePerTimeUnit.mL_Per_s);
    settings.setSubstanceCompound("Saline");
    pulse.processAction(cfg);

    pulse.advanceTime_s(10);
    // Get the values of the data you requested at this time
    data_values = pulse.pullData();
    // And write it out to the console
    dataRequests.writeData(data_values);
    
  }

}
